syntax = "proto3";

package shardapi;

// Service for communication between shards and API node
service ShardApiService {
    // Send final activation from shard to API for token generation
    rpc SendFinalActivation(FinalActivationRequest) returns (FinalActivationResponse);
    
    // Send generated token back from API to first shard
    rpc SendToken(TokenRequest) returns (TokenResponse);
}

// Final activation data from last layer
message FinalActivationRequest {
    string nonce = 1;   
    bytes data = 2;
    int32 batch_size = 3;    
    repeated int32 shape = 4;
    string dtype = 5;   
    int32 layer_id = 6; 
    int64 timestamp = 7;
    string node_origin = 8;  
}

// Response after processing final activation
message FinalActivationResponse {
    bool success = 1;
    string message = 2;  
    int32 token_id = 3;  
}

// Token to send back to first shard for next iteration
message TokenRequest {
    string nonce = 1;   
    int32 token_id = 2; 
    int64 timestamp = 3;
    float tx_enq_prev_t = 4;
}

// Response for token reception
message TokenResponse {
    bool success = 1;  
    string message = 2;
}


message RingError {
    string nonce = 1;
    string failed_node = 2;
    string error_code = 3; 
    string error = 4;
}
